ci/kind/kind-setup.sh

```diff
@@ -517,18 +517,21 @@ function destroy_external_servers {
 
 function clean_kind {
     echo "=== Cleaning up stale kind clusters ==="
-    read -a all_kind_clusters <<< $(kind get clusters)
-    for kind_cluster_name in "${all_kind_clusters[@]}"; do
-        creationTimestamp=$(kubectl get nodes --context kind-$kind_cluster_name -o json -l node-role.kubernetes.io/control-plane | \
-        jq -r '.items[0].metadata.creationTimestamp')
-        creation=$(printUnixTimestamp "$creationTimestamp")
-        now=$(date -u '+%s')
-        diff=$((now-creation))
-        timeout=$(($UNTIL_TIME_IN_MINS*60))
-        if [[ $diff -gt $timeout ]]; then
-           echo "=== kind ${kind_cluster_name} present from more than $UNTIL_TIME_IN_MINS minutes ==="
-           kind delete cluster --name $kind_cluster_name
-        fi
+    for context in $(kubectl config get-contexts -o name | grep 'kind-'); do
+      cluster_name=$(echo $context | sed 's/^kind-//')
+      creationTimestamp=$(kubectl get nodes --context $context -o json -l node-role.kubernetes.io/control-plane | \
+      jq -r '.items[0].metadata.creationTimestamp' || true)
+      if [[ -z $creationTimestamp ]]; then
+        continue
+      fi
+      creation=$(printUnixTimestamp "$creationTimestamp")
+      now=$(date -u '+%s')
+      diff=$((now-creation))
+      timeout=$(($UNTIL_TIME_IN_MINS*60))
+      if [[ $diff -gt $timeout ]]; then
+        echo "=== Cluster ${cluster_name} is present for more than $UNTIL_TIME_IN_MINS minutes ==="
+        kind delete cluster --name $cluster_name
```

Comment:
Considering the race condition, if one job retrieves the cluster name using `kubectl config get-contexts -o name` right before the `kind delete cluster` deletes the the same cluster in another job, will `kubectl get nodes --context $context` get stuck or immediately return an error if the cluster has already been deleted?

Reply:
> 2 jobs can be triggered at the same time. We should probably be tolerant of failures. If the kubectl command (kubectl get nodes) fails or the kind command fails (kind delete cluster --name $cluster_name), we don't have to fail the script.

@antoninbas I feel we still need a lock to fully avoid race conditions and ensure the cluster information remains consistent during timestamp calculations and cluster deletion 🤔️.

Reply:
I tend to agree. Honestly I think the core issue is that we try to use kubeconfig and other sources of truth to determine which clusters to delete. We rely on some specific behavior from the tools and we have to tolerate failures. It is not very important in the grand scheme of things as long as we don't corrupt or cause jobs to fail while garbage-collecting clusters. But I do feel like a lot of pain could be avoided by maintaining our own file with a list of clusters, instead of relying on `kind` or `kubeconfig`.

Before creating a cluster: write the cluster name and creation timestamp to a file, using a file lock
```bash
: "${CLUSTER_NAME:=foobar}"
flock .clusters.lock --command "echo \"$CLUSTER_NAME $(date +%s)\" >> .clusters"
# kind create cluster --name $CLUSTER_NAME
```

When doing garbage collection, use the file as the only source of truth, and also use an exclusive file lock
```bash
(
    flock -x 200

    current_timestamp=$(date +%s)
    while IFS=' ' read -r name creationTimestamp; do
        if [[ -z "$name" || -z "$creationTimestamp" ]]; then
            continue
        fi
        # Calculate the time difference
        time_difference=$((current_timestamp - creationTimestamp))
        # Check if the creation happened more than 1 hour ago (3600 seconds)
        if (( time_difference > 3600 )); then
            echo "The creation of $name happened more than 1 hour ago."
            kind delete cluster --name "$name" || echo "Cluster could not be deleted"
        else
            echo "The creation of $name happened within the last hour."
            echo "$name $creationTimestamp" >> .clusters.swp
        fi
    done < .clusters
    mv .clusters.swp .clusters
) 200>>.clusters.lock
```

IMO, correctness is more "obvious" this way, and it's not more brittle than relying on kubeconfig.
The directory where we write `.clusters` and `.clusters.lock` need to persist across Jenkins jobs.

What do you think @XinShuYang @jainpulkit22

Reply:
> IMO, correctness is more "obvious" this way, and it's not more brittle than relying on kubeconfig. The directory where we write `.clusters` and `.clusters.lock` need to persist across Jenkins jobs.
> 
> What do you think @XinShuYang @jainpulkit22

Agree, to ensure persistence, the directory for `.clusters` and `.clusters.lock` can be set to `WORKDIR` which is consistent across different Jenkins jobs. We can add a new parameter for this path in the script, making it adaptable to different environments.

Reply:
Instead of creating a separate file named ".clusters" in the testbed, I have an alternative that we can acquire the lock on the default kubeconfig file so that the other prorcess will not be able to fetch the contexts untill the lock from one is released, so this can prevent the race condition too.
What's your take on this @antoninbas @XinShuYang

Reply:
> Instead of creating a separate file named ".clusters" in the testbed, I have an alternative that we can acquire the lock on the default kubeconfig file so that the other prorcess will not be able to fetch the contexts untill the lock from one is released, so this can prevent the race condition too. What's your take on this @antoninbas @XinShuYang

I am okay with placing this lock in the `.kube` folder, but could you verify if `THIS_DIR` is the correct path that is shared by different jobs? Additionally, I recommend using `trap` to release the lock in case the context it protects exits unexpectedly, to prevent a deadlock issue.

Reply:
okay

Reply:
This seems to be based on internal knowledge of K8s and Kind (I assume they use `config.lock`, otherwise this doesn't make sense), which I am not a fan of.

> Additionally, I recommend using trap to release the lock in case the context it protects exits unexpectedly, to prevent a deadlock issue.

This should not be required, the file will be closed if the process stops, and the lock will always be unlocked. As a matter of fact, the manual states that using unlock (`flock -u`) is very rarely needed.

Reply:
> This should not be required, the file will be closed if the process stops, and the lock will always be unlocked. As a matter of fact, the manual states that using unlock (`flock -u`) is very rarely needed.

@antoninbas Yes, thanks for correcting me. It won't cause any issues once the job exits. 
I forgot to mention another concern: the commands inside the lock runs into a long hang or gets stuck. Although we do have a global timeout set for the jenkins job, that timeout is usually quite long. This means if the locked section stalls, other jobs have to wait for that long timeout period.
Do you think it‘s beneficial to set a shorter timeout specifically for the locked section, so that if it doesn’t complete in a reasonable time, we can release the lock and exit the job earlier to let other jobs proceed?

Reply:
We could have a timeout (using `flock -w`) and if we fail to acquire the lock within a reasonable time, we can skip the cleanup? Does that help?

I don't know if any of the commands are likely to stall in the locked section (once we acquire the lock). Rather than make the code too complicated, maybe we could see if any issue occurs in practice?